r/src/dataset.cpp

```diff
@@ -20,116 +20,98 @@
 #if defined(ARROW_R_WITH_ARROW)
 
 // [[arrow::export]]
-std::shared_ptr<arrow::dataset::DataSourceDiscovery> dataset___FSDSDiscovery__Make(
-    const std::shared_ptr<arrow::fs::FileSystem>& fs,
-    const std::shared_ptr<arrow::fs::Selector>& selector) {
-  std::shared_ptr<arrow::dataset::DataSourceDiscovery> discovery;
+ds::DataSourceDiscoveryPtr dataset___FSDSDiscovery__Make(
+    const std::shared_ptr<fs::FileSystem>& fs,
+    const std::shared_ptr<fs::Selector>& selector) {
   // TODO(npr): add format as an argument, don't hard-code Parquet
-  auto format = std::make_shared<arrow::dataset::ParquetFileFormat>();
+  auto format = std::make_shared<ds::ParquetFileFormat>();
   // TODO(fsaintjacques): Make options configurable
-  auto options = arrow::dataset::FileSystemDiscoveryOptions{};
+  auto options = ds::FileSystemDiscoveryOptions{};
 
-  STOP_IF_NOT_OK(arrow::dataset::FileSystemDataSourceDiscovery::Make(
-      fs.get(), *selector, format, std::move(options), &discovery));
-  return discovery;
+  return ValueOrStop(
+      ds::FileSystemDataSourceDiscovery::Make(fs, *selector, format, options));
 }
 
 // [[arrow::export]]
-std::shared_ptr<arrow::dataset::DataSource> dataset___DSDiscovery__Finish(
-    const std::shared_ptr<arrow::dataset::DataSourceDiscovery>& discovery) {
-  std::shared_ptr<arrow::dataset::DataSource> out;
-  STOP_IF_NOT_OK(discovery->Finish(&out));
-  return out;
+ds::DataSourcePtr dataset___DSDiscovery__Finish(
+    const ds::DataSourceDiscoveryPtr& discovery) {
+  return ValueOrStop(discovery->Finish());
```

Comment:
Observation: it looks odd/inconsistent to have `STOP_IF_NOT_OK` but then `ValueOrStop`

Reply:
`StopIfNoValue`?

Reply:
I meant more the capitalization and underscores

Reply:
Ah, that's because `ValueOrStop` is a function while `STOP_IF_NOT_OK` is a macro. The distinction is significant enough that in C++ we intentionally use different naming style to make it obvious which things are macros

Reply:
I understand, just observing that when it comes time to use them here, it doesn't look right.

Reply:
I'll create a trivial macro `VALUE_OR_STOP` for consistency. I also found a serious bug in STOP_IF_NOT_OK that I'll fix.

Reply:
😱 great, thank you